src/include/migraphx/op/unique.hpp

```diff
@@ -0,0 +1,321 @@
+        y_indices.resize(uniq_val_map.size());
+        y_count.resize(uniq_val_map.size());
+        size_t idx = 0;
+        // the unique elements are now sorted:
+        // post-processing for all the return indices.
+        for(const auto& v : uniq_val_map)
+        {
+            y2x_indices[v.second.y_idx] = idx;
+            y_indices[idx]              = v.second.x_idx;
+            y_count[idx]                = v.second.ct;
+            idx++;
+        }
+        // update x_rev_indices as per the sorted order of y_indices
+        for(auto& i : x_rev_indices)
+            i = y2x_indices[i];
+
+        return rv;
+    }
+
+    // CASE UNSORTED:
+    //
+    // To process into an un-sorted unique series of elements/chunks:
+    // For chunk size = 1 is a simple element, else use a flat representation of a tensor obj
+    // Go through the input elements/chunks one by one with inline processing of indices..
+
+    // INPUT x: [2, 1, 1, 3, 4, 3], attr_sorted = 0;
+
+    // OUTPUT(s): indices..
+    // y_indices: [0, 1, 3, 4]  --- first incidence, in terms of index in sequence x
+    // x_rev_indices: [0, 1, 1, 2, 3, 2] --- x seen in terms of indices of unique sequence y
+    // y_count: [1, 2, 2, 1] -- count at each y_index. sum = len(x)
+
+    // NOTE: y [2, 1, 3, 4]   --- the unique output is constructed from x[y_indices[...]]
+    // Output data structures: y_indices, x_rev_indices, y_count are processed inline.
+
+    template <class T>
+    auto unsorted_uniq_indices(const T& input_data, size_t chunk_sz) const
+    {
+        auto idx_less_fn = [&](auto idx1, auto idx2) {
+            return std::lexicographical_compare(input_data.begin() + idx1,
+                                                input_data.begin() + idx1 + chunk_sz,
+                                                input_data.begin() + idx2,
+                                                input_data.begin() + idx2 + chunk_sz);
+        };
+        std::map<size_t, size_t, decltype(idx_less_fn)> uniq_val_map(idx_less_fn);
+
+        // rv is used for NVRO below..
+        std::tuple<std::vector<std::size_t>, std::vector<std::size_t>, std::vector<std::size_t>> rv;
+        auto& [y_indices, x_rev_indices, y_count] = rv;
+
+        // go through all the elements and add the unique elements into the map..
+        // inline processing for outputs: y_indices, x_rev_indices, y_count
+        size_t count_x = input_data.size();
+        for(size_t f_idx = 0; f_idx < count_x; f_idx += chunk_sz)
+        {
+            auto [itr, added_new] = uniq_val_map.insert({f_idx, y_indices.size()});
+            if(added_new)
+            {
+                y_count.push_back(0);
+                y_indices.push_back(x_rev_indices.size());
+            }
+            y_count[itr->second]++;
+            x_rev_indices.push_back(itr->second);
+        }
+
+        return rv;
+    }
+
+    // Axis. Default: none. Range: [-rank, rank-1]
```

Comment:
So there is no normalization done for this, which could break optimization passes which will assume it is normalized. I dont think you need to enable normalization in this PR(since there are no graph transformation with `unique`), but we should have an issue open to track this.

Alternatively, we could also use `bool use_axis_0;` since we only support axis 0, and then in the future when we add support for different axes we can change this to `std::optional<int64_t> axis`.

Reply:
It already is  std::optional<int64_t> axis.

Reply:
Yes, but no normalization is done on it to remove the negative axis so we need to open an issue to fix that or bypass it completely by changing it to a simple `bool`.

Reply:
[Line 221. ](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/8cdf4aad321700d006312f2ad8852e27a130635d/src/include/migraphx/op/unique.hpp#L221)

Reply:
That only normalizes it internally for `compute_shape`, the `axis` attribute is never updated(and `compute_shape` is not the appropriate place to update it), so any optimization passes that reads the `axis` attribute will read a potential negative axis. This can lead to bugs like in #2465.

Instead, we have a normalization process that will normalize operators' fields so such consumers(like optimization passes) dont need to tune the axis. This is done in `normalize_attributes` function(which is used by the `normalize_ops` pass and `normalized_operator()` method on instructions).  This can be enabled on an operator by providing `attributes()` with which fields you want normalized and how you want it normalized. You can look at something like `concat` to see how it is setup. Probably you can use something like `value::array{normalize_attribute::include_min}` as it is done in concat.

However, I am not asking to do that in this PR, since there is no optimization pass that currently looks for `unique`, but there could be in the future so we should open an issue to address it later.

Reply:
We just validate it in once in compute_shape, and from then onwards use axis=0, the only valid choice if axis were specified. Thanks.

Reply:
> We just validate it in once in compute_shape, and from then onwards use axis=0,

Yes in `compute_shape`, but if an optimization pass(which are not part of this class) needs to read it in the future it can read either `-rank` or `0`. The optimization passes should not be tuning the axis, rather it should all be normalized to `0`.

> the only valid choice if axis were specified.

If thats the case, then it should only be the only value that can be specified, which is why I was suggesting to use `bool use_axis_0`, which makes invalid choices unrepresentable.

Reply:
OK. We should address it in another PR for the optimization pass. Thanks.